core/arch/mangle_shared.c

```diff
@@ -741,9 +741,15 @@ mangle_syscall_code(dcontext_t *dcontext, fragment_t *f, byte *pc, bool skip)
                 /* For A32 it's not OP_b_short */
                 IF_ARM(||
                        (instr_get_opcode(&instr) == OP_jmp &&
-                        opnd_get_pc(instr_get_target(&instr)) == pc + ARM_INSTR_SIZE)))
```

Comment:
I haven't verified, but could this condition be a bug for A32? It matches only when the jmp instr at `skip_pc` has `cti_pc` as target. But, it won't match when the instr at `skip_pc` has already been modified to point to the syscall.

Reply:
In this case, `skip_pc` will remain NULL, and `mangle_syscall_code` will return early. This may lead to unbounded signal delivery time.

Reply:
It looks like it will match when `skip_pc` points at the syscall (`pc` is the end of the cur instr, which is the exit cti, plus one more), so it will work for the `skip == true` case.  So it will go back to dispatch.  But it will never be re-linked and so we'll always exit it looks like: so it seems like a performance bug.

Reply:
Maybe you could tweak it to fix the problem?

Reply:
Done.

Reply:
Our test suite only builds ARM32, but doesn't run tests (https://github.com/DynamoRIO/dynamorio/wiki/Test-Suite#continuous-integration-ci-via-travis-and-appveyor). Should I do a local run of the ARM32 test suite on the Raspberry Pi machine we have (https://github.com/DynamoRIO/dynamorio/wiki/Test-Suite#ci-for-aarch64--aarch32)?

Reply:
Yes, that would be nice.  Yeah, this is a gap in the CI.  IIRC the ARM folks were going to set Jenkins up on that machine but it is not there today.  There are some academics targeting Android on 32-bit ARM who I was hoping would help set sthg up but that hasn't happened either.

Reply:
Details of manual ARM32 testing in other comment threads.